Make RenderSVGResourceGradient lazily recompute gradients when referencing element is changed

heycam opened the pull request:
#### 4f72aae07eff92f9633061e2c0bad2c73282090f
<pre>
Make RenderSVGResourceGradient lazily recompute gradients when referencing element is changed
<a href="https://bugs.webkit.org/show_bug.cgi?id=242421">https://bugs.webkit.org/show_bug.cgi?id=242421</a>
&lt;rdar://problem/96570524&gt;

Reviewed by Said Abou-Hallawa.

When an SVG element that has a gradient in its fill or stroke property
is relaid out, we call SVGResourcesCache::clientLayoutChanged to remove
and then rebuild the entire SVGResources for the element. This is
wasteful in many situations, where the specific thing that changed
during layout does not affect anything in the GradientData object that
we have stored in RenderSVGResourceGradient::m_gradientMap.

This patch changes SVGResourcesCache so that it stores any dependencies
it requires from the client RenderElement, and lazily invalidates
anything stored in GradientData under
RenderSVGResourceGradient::applyResource if those dependencies have
changed. This allows clientLayoutChanged to skip the removeClientFromCache
call if the only resources being used by the element are gradients.

Although the removeClientFromCache call which we&apos;re skipping would
repaint the client element, all callers of clientLayoutChanged already repaint
the element, so it&apos;s not necessary to add any additional repaint calls.

The new hasResourcesRequiringRemovalOnClientLayoutChange function checks
for all the remaining resource types that rely on the current
removeClientFromCache call to invalidate their cached data. The two
resource types that aren&apos;t mentioned here are markers and solid colors,
neither of which cache any data.

One of the dependencies gradients have, when painting text (on platforms
using CoreGraphics), is the AffineTransform produced by
RenderSVGResourceContainer::shouldTransformOnTextPainting. Since it&apos;s
less data to store, we change this to return the scale factor rather
than modify an existing AffineTransform.

* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::RenderSVGResourceContainer::computeTextPaintingScale):
(WebCore::RenderSVGResourceContainer::shouldTransformOnTextPainting): Deleted.
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.cpp:
(WebCore::RenderSVGResourceGradient::computeInputs):
(WebCore::RenderSVGResourceGradient::applyResource):
(WebCore::RenderSVGResourceGradient::postApplyResource):
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.h:
(WebCore::GradientData::Inputs::operator==):
(WebCore::GradientData::validate):
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
(WebCore::RenderSVGResourcePattern::buildPattern):
* Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
(WebCore::hasPaintResourceRequiringRemovalOnClientLayoutChange):
(WebCore::hasResourcesRequiringRemovalOnClientLayoutChange):
(WebCore::SVGResourcesCache::clientLayoutChanged):

Canonical link: <a href="https://commits.webkit.org/252268@main">https://commits.webkit.org/252268@main</a>
</pre>

shallawa commented on Source/WebCore/rendering/svg/RenderSVGResourceGradient.h:
```diff
@@ -31,6 +31,33 @@ namespace WebCore {
+        bool operator==(const Inputs& other)
+        {
+            return objectBoundingBox == other.objectBoundingBox && textPaintingScale == other.textPaintingScale;
+        }
```
Do we need this operator? I do not see it used. But I see we use the default operator=!() below:

`if (this->inputs != inputs) {`

heycam replied:
You are right, no need for this.

heycam replied:
Actually I am confused. Removing the `operator==` makes this line not compile.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGResourceGradient.cpp:
```diff
@@ -126,45 +139,42 @@ bool RenderSVGResourceGradient::applyResource(RenderElement& renderer, const Ren
+    if (inputs.objectBoundingBox && inputs.objectBoundingBox->isEmpty())
         return false;
 
-    bool isPaintingText = resourceMode.contains(RenderSVGResourceMode::ApplyToText);
```
Maybe it is clearer to leave this local since it is used more than once below.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGResourceGradient.h:
```diff
@@ -31,6 +31,33 @@ namespace WebCore {
+            gradient = nullptr;
+            this->inputs = inputs;
+        }
+        return gradient;
```
What is the meaning of this return value? I would expect it will return whether the object was changed or not. Assume this sequence of calls:

```
auto data = makeUnique<GradientData>();
auto inputs = computeInputs(renderer, resourceMode);
auto v1 = data->validate(inputs);
auto v2 = data->validate(inputs);
```

Both v1 and v2 will be false although the object data was changed with v1 and it did not change with v2.

heycam replied:
I was thinking that the validate function returns whether the object is valid. But maybe the inverse would be better: name this function "invalidate" (or "invalidateIfNeeded"?), and return whether it did invalidate it.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGResourceGradient.h:
```diff
@@ -31,6 +31,33 @@ namespace WebCore {
+    bool validate(const Inputs& inputs)
+    {
+        if (this->inputs != inputs) {
+            gradient = nullptr;
```
I think we should clear userspaceTransform here as well.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGResourceGradient.cpp:
```diff
@@ -126,45 +139,42 @@ bool RenderSVGResourceGradient::applyResource(RenderElement& renderer, const Ren
-        AffineTransform userspaceTransform;
+    if (!gradientData.validate(inputs)) {
+        gradientData.gradient = buildGradient(style);
+        gradientData.userspaceTransform = AffineTransform();
```
I think clearing userspaceTransform should be in GradientData::validate().

shallawa commented on Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
```diff
@@ -79,6 +79,16 @@ SVGResources* SVGResourcesCache::cachedResourcesForRenderer(const RenderElement&
+
+static bool hasResourcesRequiringRemovalOnClientLayoutChange(SVGResources& resources)
+{
+    return resources.clipper() || resources.masker() || resources.filter() || hasPaintResourceRequiringRemovalOnClientLayoutChange(resources.fill()) || hasPaintResourceRequiringRemovalOnClientLayoutChange(resources.stroke());
```
What should happen if SVGResources has a linked resource and this linked resource is a gradient?

`<linearGradient id="gradient2" href="#gradient1"/>`

heycam replied:
I'm adding a couple of tests for invalidation with linked resources in https://bugs.webkit.org/show_bug.cgi?id=242497.

heycam committed:
Make RenderSVGResourceGradient lazily recompute gradients when referencing element is changed
https://bugs.webkit.org/show_bug.cgi?id=242421
<rdar://problem/96570524>

Reviewed by Said Abou-Hallawa.

When an SVG element that has a gradient in its fill or stroke property
is relaid out, we call SVGResourcesCache::clientLayoutChanged to remove
and then rebuild the entire SVGResources for the element. This is
wasteful in many situations, where the specific thing that changed
during layout does not affect anything in the GradientData object that
we have stored in RenderSVGResourceGradient::m_gradientMap.

This patch changes SVGResourcesCache so that it stores any dependencies
it requires from the client RenderElement, and lazily invalidates
anything stored in GradientData under
RenderSVGResourceGradient::applyResource if those dependencies have
changed. This allows clientLayoutChanged to skip the removeClientFromCache
call if the only resources being used by the element are gradients.

Although the removeClientFromCache call which we're skipping would
repaint the client element, all callers of clientLayoutChanged already repaint
the element, so it's not necessary to add any additional repaint calls.

The new hasResourcesRequiringRemovalOnClientLayoutChange function checks
for all the remaining resource types that rely on the current
removeClientFromCache call to invalidate their cached data. The two
resource types that aren't mentioned here are markers and solid colors,
neither of which cache any data.

One of the dependencies gradients have, when painting text (on platforms
using CoreGraphics), is the AffineTransform produced by
RenderSVGResourceContainer::shouldTransformOnTextPainting. Since it's
less data to store, we change this to return the scale factor rather
than modify an existing AffineTransform.

* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::RenderSVGResourceContainer::computeTextPaintingScale):
(WebCore::RenderSVGResourceContainer::shouldTransformOnTextPainting): Deleted.
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.cpp:
(WebCore::RenderSVGResourceGradient::computeInputs):
(WebCore::RenderSVGResourceGradient::applyResource):
(WebCore::RenderSVGResourceGradient::postApplyResource):
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.h:
(WebCore::GradientData::Inputs::operator==):
(WebCore::GradientData::validate):
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
(WebCore::RenderSVGResourcePattern::buildPattern):
* Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
(WebCore::hasPaintResourceRequiringRemovalOnClientLayoutChange):
(WebCore::hasResourcesRequiringRemovalOnClientLayoutChange):
(WebCore::SVGResourcesCache::clientLayoutChanged):

Canonical link: https://commits.webkit.org/252268@main
